File: src/TaxAuthorityProvider/TaxAuthorities/Slovenia/ReportGeneration/IFI/CSV_D_IFI.py
from typing import Sequence

import pandas as pd

import TaxAuthorityProvider.TaxAuthorities.Slovenia.Schemas.Schemas as ss


# NOTE: When comparing with exports from IBKR, take the realized P/L and add comissions. EDavki does reporting based on Trade Price, not Cost Basis !!!
# The generated reports are going to show you made more money than you really did because Slovenia recognizes 1% of the Trade Price as the costs associated with buying/selling of the underlying.
def generateDataFrameReport(convertedTrades: Sequence[ss.EDavkiGenericDerivativeReportItem]) -> pd.DataFrame:
    def getLinesFromData(
        entry: ss.EDavkiGenericDerivativeReportItem,
    ) -> pd.DataFrame:

        lines = pd.DataFrame(entry.Items)
        lines["ISIN"] = entry.ISIN
        lines["Ticker"] = entry.Code
        lines["Name"] = entry.Name
        lines["HasForeignTax"] = entry.HasForeignTax
        lines["ForeignTax"] = entry.ForeignTax
        lines["ForeignTaxCountryID"] = entry.FTCountryID
        lines["ForeignTaxCountryName"] = entry.FTCountryName

        return lines

    mappedData = list(map(getLinesFromData, convertedTrades))

    if len(mappedData) == 0:
        return pd.DataFrame()

    combinedData = pd.concat(mappedData)

    combinedData.insert(0, "Name", combinedData.pop("Name"))

    return combinedData

File: lib/src/TaxAuthorities/Slovenia/ReportGeneration/Ifi/CsvDIfi.ts
import {
	EDavkiDerivativeReportSecurityLineGenericEventBought,
	type EDavkiGenericDerivativeReportItem,
} from "@brrr/TaxAuthorities/Slovenia/Schemas/Schemas";
import { stringify } from "csv-stringify/sync";

export function generateCsvReport(convertedTrades: EDavkiGenericDerivativeReportItem[]): string {
	if (convertedTrades.length === 0) {
		return "";
	}

	const rows = [];

	for (const entry of convertedTrades) {
		for (const item of entry.items) {
			const isBuy = item instanceof EDavkiDerivativeReportSecurityLineGenericEventBought;
			const row = {
				Name: entry.name,
				BoughtOn: isBuy ? item.boughtOn.toISO() : null,
				GainType: isBuy ? item.gainType : null,
				SoldOn: isBuy ? null : item.soldOn.toISO(),
				Quantity: item.quantity,
				PricePerUnit: item.pricePerUnit,
				PricePerUnitInOriginalCurrency: item.pricePerUnitInOriginalCurrency,
				TotalPrice: item.totalPrice,
				TotalPriceInOriginalCurrency: item.totalPriceInOriginalCurrency,
				Commissions: item.commissions,
				CommissionsInOriginalCurrency: item.commissionsInOriginalCurrency,
				Leveraged: item.leveraged,
				ISIN: entry.isin,
				Ticker: entry.code,
				HasForeignTax: entry.hasForeignTax,
				ForeignTax: entry.foreignTax,
				ForeignTaxCountryID: entry.ftCountryId,
				ForeignTaxCountryName: entry.ftCountryName,
			};
			rows.push(row);
		}
	}

	return stringify(rows, { header: true, cast: { boolean: (v) => String(v) } });
}
